### PR TITLE
Order code actions

### PR DESCRIPTION
**Pull request by akshita31**
Fixes : #758 

To order code actions we can order the providers based on the ExtensionOrder attribute in roslyn.There are two types of providers - CodeFixProvider and CodeRefactoringProvider. For each of these categories we perform the following on a list of providers : 

1. For each provider create a ProviderNode that includes the Before and After strings and the ProviderName
2. Create a dictionary that includes a mapping from the provider name to the providernode
3. Create a graph based on the Before and After properties of the node using the dictionary created above
4. Perform topological sorting on the nodes. 

Please review @DustinCampbell @TheRealPiotrP @rchande @filipw

### CODE REVIEW

File: src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/CodeActionsOrder.Graph.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace OmniSharp.Roslyn.CSharp.Services.Refactoring.V2
+{
+    internal class Graph<T>
```
Comment (rchande) [inline src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/CodeActionsOrder.Graph.cs]: Is this adapted from Roslyn?

File: src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/BaseCodeActionService.cs [hunk 2]
```diff
@@ -128,26 +125,54 @@ private TextSpan GetTextSpan(ICodeActionRequest request, SourceText sourceText)
...
+            foreach (var provider in this.Providers)
+            {
+                foreach (var codeFixProvider in provider.CodeFixProviders)
+                {
+                    nodesList.Add(ProviderNode<CodeFixProvider>.From(codeFixProvider));
                 }
             }
+
+            var graph = Graph<CodeFixProvider>.GetGraph(nodesList);
+            return graph.TopologicalSort();
```
Comment (rchande) [inline src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/BaseCodeActionService.cs]: Can we cache this? (Eg, does the total set of CodeFixProviders change)

File: src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/CodeActionsOrder.Graph.cs (hunk 1 above)
Comment (akshita31) [reply src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/CodeActionsOrder.Graph.cs]: Yes, this is taken from [Extension Orderer](https://github.com/dotnet/roslyn/blob/614299ff83da9959fa07131c6d0ffbc58873b6ae/src/Workspaces/Core/Portable/Shared/Utilities/ExtensionOrderer.cs) and modified a little for the current case.

Comment (rchande) [reply src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/CodeActionsOrder.Graph.cs]: Ok, we might want to add a comment stating as such.

Comment (filipw) [reply src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/CodeActionsOrder.Graph.cs]: @DustinCampbell can we get that API opened up publicly? 😀

File: src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/BaseCodeActionService.cs [hunk 3]
```diff
@@ -23,6 +23,8 @@ public abstract class BaseCodeActionService<TRequest, TResponse> : IRequestHandl
     {
         protected readonly OmniSharpWorkspace Workspace;
         protected readonly IEnumerable<ICodeActionProvider> Providers;
+        protected List<CodeFixProvider> OrderedCodeFixProviders;
```
Comment (rchande) [inline src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/BaseCodeActionService.cs]: You can make this a `Lazy<IEnumerable<CodeFixProvider>` and initialize it like
```
_lazyFixProviders = new Lazy(() => GetOrderedProviders)```

That will let you avoid some of the crazy exception handling below.

File: src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/BaseCodeActionService.cs [hunk 4]
```diff
@@ -51,6 +55,41 @@ protected BaseCodeActionService(OmniSharpWorkspace workspace, CodeActionHelper h
             }
         }
 
+        private void SetOrderedProviders()
+        {
+            try
+            {
```
Comment (rchande) [inline src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/BaseCodeActionService.cs]: Why is this in a try/catch? What are you expecting to fail?

Comment (akshita31) [reply src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/BaseCodeActionService.cs]: the GetSortedCodeFixProviders() will return an exception if there are cycles in the graph. This try catch will handle the exception and return the unordered list if that exception occurs

**Comment by filipw** [comment]
green!

**Pull request merged**